### PR TITLE
Allow multiple segments in DockerImageReference

### PR DESCRIPTION
**Pull request by legionus**
- [x] add an integration tests
- [x] profit

@mfojtik @liggitt @smarterclayton PTAL

### CODE REVIEW

**Comment by legionus** [comment]
@miminar PTAL

File: pkg/image/api/helper.go [hunk 1]
```diff
@@ -191,6 +178,61 @@ func ParseDockerImageReference(spec string) (DockerImageReference, error) {
...
+	ref.ID = id
+
+	return ref, err
+}
+
+func ParseMultiSegmentsDockerImageReference(spec string) (DockerImageReference, error) {
+	stream, tag, id := parseRepositoryTag(spec)
+
+	repoParts := strings.Split(stream, "/")
+	if len(repoParts) < 4 {
```
Comment (liggitt) [inline pkg/image/api/helper.go]: explain the magic 4 in a comment?

File: pkg/image/api/helper.go [hunk 2]
```diff
@@ -191,6 +178,61 @@ func ParseDockerImageReference(spec string) (DockerImageReference, error) {
...
+	ref.Registry = repoParts[0]
+
+	if len(repoParts[1]) == 0 {
+		return ref, fmt.Errorf("the docker pull spec %q must contain non-empty namespace", spec)
+	}
+	ref.Namespace = repoParts[1]
+
+	ref.Name = strings.Join(repoParts[2:], "/")
+	if len(ref.Name) == 0 {
+		return ref, fmt.Errorf("the docker pull spec %q must be more than three segments separated by slashes", spec)
```
Comment (liggitt) [inline pkg/image/api/helper.go]: "more than three" or "three or more"?

**Comment by legionus** [comment]
[test]

File: pkg/image/api/helper_test.go [hunk 3]
```diff
@@ -122,10 +122,16 @@ func TestParseImageStreamTagName(t *testing.T) {
...
+	parserFuncs := map[string]func(spec string) (DockerImageReference, error){
+		"ParseDockerImageReference":              ParseDockerImageReference,
+		"ParseMultiSegmentsDockerImageReference": ParseMultiSegmentsDockerImageReference,
+	}
+
 	testCases := []struct {
 		From                               string
 		Registry, Namespace, Name, Tag, ID string
 		Err                                bool
+		MSegErr                            bool
```
Comment (liggitt) [inline pkg/image/api/helper_test.go]: this is hard to follow, just duplicate the cases and make the parsing function one of the inputs to the test

File: pkg/image/api/helper.go [hunk 4]
```diff
@@ -191,6 +178,61 @@ func ParseDockerImageReference(spec string) (DockerImageReference, error) {
 	return ref, nil
 }
 
+// ParseDockerImageReference parses a Docker pull spec string into a
+// DockerImageReference.
+func ParseDockerImageReference(spec string) (DockerImageReference, error) {
+	// TODO replace with docker version once docker/docker PR11109 is merged upstream
```
Comment (miminar) [inline pkg/image/api/helper.go]: I think it's about time to resolve this. docker/docker#11109 was marched in March 2015.

File: pkg/image/api/helper.go [hunk 5]
```diff
@@ -191,6 +178,61 @@ func ParseDockerImageReference(spec string) (DockerImageReference, error) {
...
+		ref.ID = id
+
+		return ref, err
+	}
+
+	var ref DockerImageReference
+
+	// registry/namespace/name/namecomponent/namecomponent/...
+	if strings.HasSuffix(repoParts[0], ":") || IsRegistryDockerHub(repoParts[0]) {
+		return ref, fmt.Errorf("the docker pull spec %q must be two or three segments separated by slashes", spec)
```
Comment (miminar) [inline pkg/image/api/helper.go]: The error message looks outdated.

File: pkg/image/api/helper.go [hunk 6]
```diff
@@ -130,14 +130,9 @@ func IsRegistryDockerHub(registry string) bool {
 	}
 }
 
-// ParseDockerImageReference parses a Docker pull spec string into a
-// DockerImageReference.
-func ParseDockerImageReference(spec string) (DockerImageReference, error) {
+func parseDockerImageReference(repoParts []string, spec string) (DockerImageReference, error) {
```
Comment (miminar) [inline pkg/image/api/helper.go]: Having two different functions with the same name looks suspicious or at least redundant.

File: pkg/image/api/helper.go (hunk 1 above)
Comment (legionus) [inline pkg/image/api/helper.go]: Right now we can have maximum 3 part (`registry/namespace/name`). Multi-segment path means  4 or more. So, old behavior will be in case of < 4.

File: pkg/image/api/helper.go [hunk 7]
```diff
@@ -191,6 +178,61 @@ func ParseDockerImageReference(spec string) (DockerImageReference, error) {
...
+
+	var ref DockerImageReference
+
+	// registry/namespace/name/namecomponent/namecomponent/...
+	if strings.HasSuffix(repoParts[0], ":") || IsRegistryDockerHub(repoParts[0]) {
+		return ref, fmt.Errorf("the docker pull spec %q must be two or three segments separated by slashes", spec)
+	}
+	ref.Registry = repoParts[0]
+
+	if len(repoParts[1]) == 0 {
```
Comment (miminar) [inline pkg/image/api/helper.go]: Can this be applied to all the name components?

File: pkg/image/api/helper_test.go (hunk 3 above)
Comment (legionus) [inline pkg/image/api/helper_test.go]: I wanted to make sure that all the tests pass by both functions except for some testcases and wanted to quickly find out which tests are different.

File: pkg/image/api/helper.go (hunk 4 above)
Comment (legionus) [inline pkg/image/api/helper.go]: Yes, but the parser is very different from what we do. I tried it.

File: pkg/image/api/validation/validation.go [hunk 8]
```diff
@@ -56,7 +56,7 @@ func validateImage(image *api.Image, fldPath *field.Path) field.ErrorList {
 	if len(image.DockerImageReference) == 0 {
 		result = append(result, field.Required(fldPath.Child("dockerImageReference"), ""))
 	} else {
-		if _, err := api.ParseDockerImageReference(image.DockerImageReference); err != nil {
+		if _, err := api.ParseMultiSegmentsDockerImageReference(image.DockerImageReference); err != nil {
```
Comment (miminar) [inline pkg/image/api/validation/validation.go]: What about parsing docker image reference in `ValidateImageStream()`.

File: pkg/image/api/helper.go (hunk 5 above)
Comment (legionus) [inline pkg/image/api/helper.go]: No. This is the error message that was returned by `ParseDockerImageReference` in same case.

File: pkg/image/api/helper.go (hunk 6 above)
Comment (legionus) [inline pkg/image/api/helper.go]: OK. I will change it to something more common.

File: pkg/image/api/helper.go (hunk 7 above)
Comment (legionus) [inline pkg/image/api/helper.go]: @miminar I'm not sure I understand.

File: pkg/image/api/validation/validation.go (hunk 8 above)
Comment (legionus) [inline pkg/image/api/validation/validation.go]: @miminar  `ValidateImageStream` must be unmodified.

Comment (miminar) [inline pkg/image/api/validation/validation.go]: Also this is a good candidate for a new unit test.

File: pkg/build/api/validation/validation.go [hunk 9]
```diff
@@ -394,7 +394,7 @@ func validateFromImageReference(reference *kapi.ObjectReference, fldPath *field.
 		}
 		if len(name) == 0 {
 			allErrs = append(allErrs, field.Required(fldPath.Child("name"), ""))
-		} else if _, err := imageapi.ParseDockerImageReference(name); err != nil {
+		} else if _, err := imageapi.ParseMultiSegmentsDockerImageReference(name); err != nil {
```
Comment (miminar) [inline pkg/build/api/validation/validation.go]: It seems to me like we need to update `validateToImageReference` as well.

File: pkg/image/api/helper.go (hunk 4 above)
Comment (miminar) [inline pkg/image/api/helper.go]: If you think it's not worth it, just delete the comment. On the other hand, I'd like to have a way to easily turn our `DockerImageReference` into upstream's `reference.Reference`.

File: pkg/build/api/validation/validation.go (hunk 9 above)
Comment (legionus) [inline pkg/build/api/validation/validation.go]: @miminar I was not sure about that. I will fix it.

File: pkg/image/api/helper.go (hunk 5 above)
Comment (miminar) [inline pkg/image/api/helper.go]: The error message contradicts a comment two lines above.

File: pkg/image/api/helper.go (hunk 7 above)
Comment (miminar) [inline pkg/image/api/helper.go]: I mean all the name components shall be not empty. Why not checking them all in a loop?

File: pkg/image/api/helper.go (hunk 4 above)
Comment (legionus) [inline pkg/image/api/helper.go]: @miminar No, I mean that must be a completely different PR because we have some additional magic guessing (https://github.com/openshift/origin/blob/master/pkg/image/api/helper.go#L142-L156) and restriction how many components we can accept. So, we can't use `reference.Reference` without extra work anyway.

File: pkg/image/api/helper.go (hunk 5 above)
Comment (legionus) [inline pkg/image/api/helper.go]: Will be OK if I put comment `// http://registry/namespace/name` inside `if` ?

Comment (miminar) [inline pkg/image/api/helper.go]: I don't quite understand, why a function that is supposed to parse a reference with way more than 3 segments would ever return `"the docker pull spec %q must be two or three segments separated by slashes"`. The error statement just isn't right.

File: pkg/image/api/validation/validation.go (hunk 8 above)
Comment (legionus) [inline pkg/image/api/validation/validation.go]: @miminar I'll do more tests when this approach will be approved by @liggitt, and @smarterclayton and you, because I'm not sure that we want to make these changes.

File: pkg/image/api/helper.go (hunk 7 above)
Comment (legionus) [inline pkg/image/api/helper.go]: Yes, I miss check for `ref.Registry`. Thanks! I'll fix it.

**Comment by miminar** [comment]
Could this add an integration test which would launch a fake registry server serving manifests from multi-slash repositories which we would import?

File: pkg/image/api/helper.go (hunk 5 above)
Comment (legionus) [inline pkg/image/api/helper.go]: @miminar because this case for `http://example.com/namespace/name`. 

But I just realized that I can't act like that and in case of specifying of schema I need to parse the rest path. Thanks!

Comment (legionus) [inline pkg/image/api/helper.go]: @miminar I was wrong again. We don't support schema in `DockerImageReference`:
https://github.com/openshift/origin/blob/master/pkg/image/api/helper_test.go#L290-L291
https://github.com/openshift/origin/blob/master/pkg/image/api/helper_test.go#L298-L299

File: pkg/image/api/helper.go (hunk 4 above)
Comment (smarterclayton) [inline pkg/image/api/helper.go]: Please document the special behavior in the godoc for these methods - anything this fundamental should be documented well.  Please include the official spec, and each point where we differ.

File: pkg/image/api/helper_test.go (hunk 3 above)
Comment (smarterclayton) [inline pkg/image/api/helper_test.go]: I agree with Jordan - more explicit tests are better.  Use a common core array and add variant cases.

File: pkg/image/api/helper.go [hunk 10]
```diff
@@ -191,6 +178,61 @@ func ParseDockerImageReference(spec string) (DockerImageReference, error) {
...
+	if err != nil {
+		return ref, err
+	}
+	ref.Tag = tag
+	ref.ID = id
+
+	return ref, err
+}
+
+func ParseMultiSegmentsDockerImageReference(spec string) (DockerImageReference, error) {
```
Comment (smarterclayton) [inline pkg/image/api/helper.go]: This should not be public - we should not have *two* DockerImageReferences in the system.  IF something is using ParseDockerImageReference that is not a valid DockerImageReference, we should create a new method which is "ParseNotQuiteDockerImageReference" (name can be better) that other code should call.  DockerImageReferences are multi segment by default, which means ParseDockerImageReference should support multi-segment.

Comment (legionus) [inline pkg/image/api/helper.go]: @smarterclayton you mean to add an argument to `ParseDockerImageReference` to allow or deny multi-segments ?

**Comment by legionus** [comment]
[test]

**Comment by legionus** [comment]
[test]

File: pkg/image/api/helper.go [hunk 11]
```diff
@@ -184,7 +184,41 @@ func ParseDockerImageReference(spec string) (DockerImageReference, error) {
 		ref.ID = id
 		break
 	default:
-		// TODO: this is no longer true with V2
+		// Handle multiple segments in form: registry/namespace/namesegment1/namesegment2/.../name
+		// but not this form: http://registry/namespace/name. We don't support using the schema in
+		// the DockerImageReference.
+		if len(repoParts) > 3 && !strings.HasSuffix(repoParts[0], ":") {
```
Comment (ncdc) [inline pkg/image/api/helper.go]: Isn't this going to miss validating myregistry.com:5000/a/b/c?

File: pkg/image/api/helper.go [hunk 12]
```diff
@@ -184,7 +184,41 @@ func ParseDockerImageReference(spec string) (DockerImageReference, error) {
...
+				// DockerHub doesn't support multiple segments.
+				return ref, fmt.Errorf("the docker pull spec %q must be two or three segments separated by slashes", spec)
+			}
+			if len(repoParts[0]) == 0 {
+				return ref, fmt.Errorf("the docker pull spec %q must contain non-empty registry name", spec)
+			}
+			ref.Registry = repoParts[0]
+
+			if len(repoParts[1]) == 0 {
+				return ref, fmt.Errorf("the docker pull spec %q must contain non-empty namespace", spec)
```
Comment (ncdc) [inline pkg/image/api/helper.go]: Technically it's not a "namespace". It's just a "component". See https://github.com/docker/distribution/blob/744ae974a5bfc47c219320460eead041fc9f64f4/reference/reference.go#L7.

File: pkg/image/api/helper.go [hunk 13]
```diff
@@ -184,7 +184,41 @@ func ParseDockerImageReference(spec string) (DockerImageReference, error) {
...
+			}
+			ref.Namespace = repoParts[1]
+
+			for i, part := range repoParts[2:] {
+				if len(part) == 0 {
+					continue
```
Comment (ncdc) [inline pkg/image/api/helper.go]: Shouldn't this error?

File: pkg/image/api/helper.go [hunk 14]
```diff
@@ -184,7 +184,41 @@ func ParseDockerImageReference(spec string) (DockerImageReference, error) {
...
+				}
+				if i > 0 {
+					ref.Name += "/"
+				}
+				ref.Name += part
+			}
+			if len(ref.Name) == 0 {
+				return ref, fmt.Errorf("the docker pull spec %q must be three or more segments separated by slashes", spec)
```
Comment (ncdc) [inline pkg/image/api/helper.go]: I don't think this error message is accurate. It does not have to be three or more segments. In fact, for all of the error messages we return from this function, wouldn't it make more sense to just say it's not a valid pull spec, and if we can, give a reason why it failed to parse (registry is missing, segment can't be empty, etc). I think it might be time to remove the error text saying it has to be some specific # of segments.

File: pkg/image/api/helper_test.go [hunk 15]
```diff
@@ -291,8 +291,16 @@ func TestParseDockerImageReference(t *testing.T) {
 			Err:  true,
 		},
 		{
-			From: "bar/foo/baz/biz",
-			Err:  true,
+			From:      "bar/foo/baz/biz",
+			Registry:  "bar",
```
Comment (ncdc) [inline pkg/image/api/helper_test.go]: Technically, I think if the docker hub ever starts supporting multiple components, this would resolve to registry=docker.io, namespace=bar, name=foo/baz/biz.

File: pkg/build/api/validation/validation_test.go [hunk 16]
```diff
@@ -1269,7 +1269,7 @@ func TestValidateCommonSpec(t *testing.T) {
 				Output: buildapi.BuildOutput{
 					To: &kapi.ObjectReference{
 						Kind: "DockerImage",
-						Name: "some/long/value/with/no/meaning",
+						Name: "index.docker.io/some/long/value/with/no/meaning",
```
Comment (ncdc) [inline pkg/build/api/validation/validation_test.go]: Why these changes?

Comment (liggitt) [inline pkg/build/api/validation/validation_test.go]: it's changing because this is now a valid image spec... only by pinning it to dockerhub can we trigger the invalid error?

Comment (ncdc) [inline pkg/build/api/validation/validation_test.go]: Ah, right, at least for the hub. Although as I wrote in another comment, if/when the hub allows `[user]/[one]/[or]/[more]/[segments]`, these tests won't be valid any more

File: pkg/image/api/helper.go (hunk 11 above)
Comment (legionus) [inline pkg/image/api/helper.go]: @ncdc No. This check only for [suffix](https://play.golang.org/p/qJAi2U4USb), not for substring.

Comment (ncdc) [inline pkg/image/api/helper.go]: Oh I see it now, thanks. Nevermind :-) But when would it have a suffix of :?

File: pkg/image/api/helper.go (hunk 12 above)
Comment (legionus) [inline pkg/image/api/helper.go]: Yes, but the text of the error made in the context of DockerImageReference.
Should I write something like:
`the docker pull spec %q must be three or more segments separated by slashes`
?

it seemed to me that it's less obvious in this case.

File: pkg/image/api/helper.go (hunk 13 above)
Comment (legionus) [inline pkg/image/api/helper.go]: OK. I will change it.

File: pkg/image/api/helper.go (hunk 12 above)
Comment (ncdc) [inline pkg/image/api/helper.go]: I would prefer something that indicates that you can't have any empty segments/components. Also see my comment below about removing any specificity of number of segments.

File: pkg/image/api/helper.go (hunk 11 above)
Comment (legionus) [inline pkg/image/api/helper.go]: `http://example.com/foo/bar`

Comment (ncdc) [inline pkg/image/api/helper.go]: That's not valid. We've never supported a scheme. Is that all you're testing the suffix for?

Comment (legionus) [inline pkg/image/api/helper.go]: Also hostname can't have `:` at the end anyway.

Comment (ncdc) [inline pkg/image/api/helper.go]: Right... I guess this is ok if we actually have users putting in `http://` at the front of their pull specs...

File: pkg/build/api/validation/validation_test.go (hunk 16 above)
Comment (legionus) [inline pkg/build/api/validation/validation_test.go]: But we use our parser, so these tests will be valid until we change it and allow mutliple segments for dockerhub. These tests we will change along with the parser.

Comment (legionus) [inline pkg/build/api/validation/validation_test.go]: The purpose of these tests is not the link itself but any invalid value will cause an error.

File: pkg/image/api/helper.go (hunk 11 above)
Comment (legionus) [inline pkg/image/api/helper.go]: @ncdc We have a tests that verify this ban:

https://github.com/openshift/origin/blob/master/pkg/image/api/helper_test.go#L290-L291

https://github.com/openshift/origin/blob/master/pkg/image/api/helper_test.go#L298-L299

File: pkg/image/api/helper.go [hunk 17]
```diff
@@ -184,7 +184,41 @@ func ParseDockerImageReference(spec string) (DockerImageReference, error) {
 		ref.ID = id
 		break
 	default:
-		// TODO: this is no longer true with V2
```
Comment (miminar) [inline pkg/image/api/helper.go]: Regarding `case 3`: I feel like it's about time to recognize `registry` part based on special characters it contains. The same way [upstream does it](https://github.com/docker/docker/blob/0cc7f0185f59f9271b7b783b776dac74347e8f51/reference/reference.go#L176).

File: pkg/image/api/helper_test.go (hunk 15 above)
Comment (miminar) [inline pkg/image/api/helper_test.go]: That's right. See my above comment https://github.com/openshift/origin/pull/11173/files#r81712056

File: pkg/image/api/validation/validation_test.go [hunk 18]
```diff
@@ -315,24 +315,6 @@ func TestValidateImageStreamMappingNotOK(t *testing.T) {
 			field.ErrorTypeRequired,
 			"image.metadata.name",
 		},
-		"invalid repository pull spec": {
```
Comment (miminar) [inline pkg/image/api/validation/validation_test.go]: Instead of removing, could you change the test to check this is a valid reference? Also could you add another case that would trigger `ErrorTypeInvalid`?

File: pkg/build/api/validation/validation_test.go (hunk 16 above)
Comment (legionus) [inline pkg/build/api/validation/validation_test.go]: @miminar try to create such repository on docker.io

File: pkg/image/api/helper.go (hunk 17 above)
Comment (legionus) [inline pkg/image/api/helper.go]: @miminar parser refactoring out of scope of PR. I can open following issue about it.

File: pkg/image/api/validation/validation_test.go (hunk 18 above)
Comment (legionus) [inline pkg/image/api/validation/validation_test.go]: @miminar This test is called the `TestValidateImageStreamMappingNotOK`. So, this case can't be here anyway.

**Comment by legionus** [comment]
@miminar @ncdc PTAL again.

File: pkg/image/api/validation/validation_test.go (hunk 18 above)
Comment (miminar) [inline pkg/image/api/validation/validation_test.go]: Ok, so at least change the repository to something like `///a/b/c/d` to generate the `ErrorTypeInvalid`.

File: pkg/image/api/helper.go (hunk 17 above)
Comment (miminar) [inline pkg/image/api/helper.go]: Ahh, I hate devcuts.

**Comment by legionus** [comment]
[test]

File: test/integration/imageimporter_test.go [hunk 19]
```diff
@@ -316,6 +316,14 @@ func TestImageStreamImportOfV1ImageFromV2Repository(t *testing.T) {
 	}
 }
 
+func TestImageStreamImportOfV1ImageFromV2Repository(t *testing.T) {
+	testImageStreamImportWithPath(t, "/test/image")
```
Comment (legionus) [inline test/integration/imageimporter_test.go]: This is a part of path.

Comment (legionus) [inline test/integration/imageimporter_test.go]: @miminar I need to rename second argument ?

File: pkg/image/api/helper.go (hunk 14 above)
Comment (ncdc) [inline pkg/image/api/helper.go]: Please please please change the error text. I do not want it saying that it has to be three or more segments.

File: pkg/image/api/helper.go [hunk 20]
```diff
@@ -184,7 +184,41 @@ func ParseDockerImageReference(spec string) (DockerImageReference, error) {
...
+		// Handle multiple segments in form: registry/namespace/namesegment1/namesegment2/.../name
+		// but not this form: http://registry/namespace/name. We don't support using the schema in
+		// the DockerImageReference.
+		if len(repoParts) > 3 && !strings.HasSuffix(repoParts[0], ":") {
+			if IsRegistryDockerHub(repoParts[0]) {
+				// DockerHub doesn't support multiple segments.
+				return ref, fmt.Errorf("the docker pull spec %q must be two or three segments separated by slashes", spec)
```
Comment (ncdc) [inline pkg/image/api/helper.go]: I would reword as "the docker pull spec %q must be of the form <username>/<repository> to be valid for the docker hub" or something like that

Comment (legionus) [inline pkg/image/api/helper.go]: OK. I will use your message.

File: pkg/image/api/helper.go (hunk 14 above)
Comment (legionus) [inline pkg/image/api/helper.go]: @ncdc Actually this `if` is useless because if have more than 3 parts in slice (`repoParts`), two of them are not empty (`Registry` and `Namespace`) and rest of them are not empty as well after check inside `for`. It means that `Name` can't be empty.
I will remove this check.

File: test/integration/imageimporter_test.go (hunk 19 above)
Comment (legionus) [inline test/integration/imageimporter_test.go]: @miminar Done. `/` removed.

File: pkg/image/api/validation/validation_test.go (hunk 18 above)
Comment (legionus) [inline pkg/image/api/validation/validation_test.go]: @miminar done

**Comment by ncdc** [comment]
I think we'll be best served by doing the following (which, yes, is a parser refactoring):

1. Use https://github.com/docker/distribution/blob/master/reference/reference.go#L150 to parse the initial pull spec string
1. Call https://github.com/docker/distribution/blob/master/reference/reference.go#L138 to get the registry hostname, if specified
1. Now that we have the reference from step 1 (and possibly a hostname from step 2):
  1. Fill in ref.Registry from hostname
  1. Extract name/tag/digest using type assertions against the distribution interfaces
  1. Perform any additional validation we require

**Comment by legionus** [comment]
@ncdc devcut is tomorrow. We don't have enough time to make such refactoring.
@mfojtik @smarterclayton Are you agree to move this [fix](https://bugzilla.redhat.com/show_bug.cgi?id=1373281) to the next release ?

**Comment by ncdc** [comment]
We should do it right. What release are you currently targeting?

**Comment by legionus** [comment]
@ncdc proposed-3.4

**Comment by legionus** [comment]
[test]

**Comment by legionus** [comment]
[test]

**Comment by ncdc** [comment]
Ok, I looked at the docker/distribution code. It doesn't cover everything we have ("case 2:" for repoParts). So we can proceed with this PR for now and hopefully switch to the upstream parser later.

File: pkg/image/api/helper.go [hunk 21]
```diff
@@ -184,7 +184,38 @@ func ParseDockerImageReference(spec string) (DockerImageReference, error) {
...
 		break
 	default:
-		// TODO: this is no longer true with V2
+		// Handle multiple segments in form: registry/namespace/namesegment1/namesegment2/.../name
+		// but not this form: http://registry/namespace/name. We don't support using the schema in
+		// the DockerImageReference.
+		if len(repoParts) > 3 && !strings.HasSuffix(repoParts[0], ":") {
+			if IsRegistryDockerHub(repoParts[0]) {
+				// DockerHub doesn't support multiple segments.
+				return ref, fmt.Errorf("the docker pull spec %q must be of the form / to be valid for the docker hub", spec)
```
Comment (ncdc) [inline pkg/image/api/helper.go]: Looks like something got lost in translation from my original suggestion. `the docker pull spec %q must be of the form <username>/<repository> to be valid for the docker hub`

File: pkg/image/api/helper.go [hunk 22]
```diff
@@ -184,7 +184,38 @@ func ParseDockerImageReference(spec string) (DockerImageReference, error) {
...
+			}
+			if len(repoParts[0]) == 0 {
+				return ref, fmt.Errorf("the docker pull spec %q must contain non-empty registry name", spec)
```
Comment (ncdc) [inline pkg/image/api/helper.go]: I would either say you can't have empty segments, or you can't have 2 consecutive slashes (`//`). I'm really trying hard to avoid us telling a user that they **have** to have something when it's not necessarily required.

Comment (ncdc) [inline pkg/image/api/helper.go]: Or you could say "registry cannot be empty"; I could live with that 😄

Comment (legionus) [inline pkg/image/api/helper.go]: But first segment here have to be registry. No? Why can't tell user that he miss and what we expect to see ?

Comment (ncdc) [inline pkg/image/api/helper.go]: If/when the hub supports multiple segments, no, the first segment doesn't have to be the registry. But today, yes, it has to be the registry. This is fine for now.

File: pkg/image/api/helper.go (hunk 21 above)
Comment (legionus) [inline pkg/image/api/helper.go]: It's not true. You can use `docker.io/busybox`. So, `<username>` not required.

Comment (ncdc) [inline pkg/image/api/helper.go]: Yeah, the fun of the implicit `library`. I'd personally prefer to remove this logic that makes a special case out of the hub, even though a hub spec with more than 2 segments isn't valid at the moment. WDYT? @liggitt @mfojtik @pweil- @smarterclayton

Comment (legionus) [inline pkg/image/api/helper.go]: @ncdc How about `invalid docker hub pull spec %q` ?

File: pkg/image/api/helper.go (hunk 22 above)
Comment (legionus) [inline pkg/image/api/helper.go]: @ncdc `the docker pull spec %q cannot have empty segments` -- it's ok for you ?

Comment (ncdc) [inline pkg/image/api/helper.go]: That's fine, thanks.

File: pkg/image/api/helper.go (hunk 21 above)
Comment (ncdc) [inline pkg/image/api/helper.go]: Either that, or remove this logic if everyone else agrees

Comment (legionus) [inline pkg/image/api/helper.go]: @ncdc I added this condition because we have previously banned more than 3 parts, and by this we banned `docker.io/foo/bar/baz` as well. I wanted to allow multiple segments only for custom `docker/distribution` servers.

Comment (ncdc) [inline pkg/image/api/helper.go]: If the hub moves to add more than 2 segments, then we'd have to refactor, rebuild, and put out a new release to support it. I'd rather avoid that if we can.

File: pkg/image/api/helper.go (hunk 22 above)
Comment (legionus) [inline pkg/image/api/helper.go]: @ncdc fixed.

File: pkg/image/api/helper.go (hunk 21 above)
Comment (legionus) [inline pkg/image/api/helper.go]: @ncdc error message changed.

Comment (legionus) [inline pkg/image/api/helper.go]: @ncdc This check has been removed.

File: pkg/image/api/helper.go [hunk 23]
```diff
@@ -184,7 +184,34 @@ func ParseDockerImageReference(spec string) (DockerImageReference, error) {
 		ref.ID = id
 		break
 	default:
-		// TODO: this is no longer true with V2
+		// Handle multiple segments in form: registry/namespace/namesegment1/namesegment2/.../name
+		// but not this form: http://registry/namespace/name. We don't support using the schema in
+		// the DockerImageReference.
+		if len(repoParts) > 3 && !strings.HasSuffix(repoParts[0], ":") {
+			if len(repoParts[0]) == 0 {
+				return ref, fmt.Errorf("the docker pull spec %q cannot have empty segments", spec)
```
Comment (ncdc) [inline pkg/image/api/helper.go]: Make this error text consistent with the 2 below please

File: pkg/image/api/helper.go [hunk 24]
```diff
@@ -184,7 +184,34 @@ func ParseDockerImageReference(spec string) (DockerImageReference, error) {
...
+					ref.Name += "/"
+				}
+				ref.Name += part
+			}
+
+			ref.Tag = tag
+			ref.ID = id
+			break
+		}
 		return ref, fmt.Errorf("the docker pull spec %q must be two or three segments separated by slashes", spec)
```
Comment (ncdc) [inline pkg/image/api/helper.go]: To get to this line of code, it means either:

1. len(repoParts) > 3 and repoparts[0] is a scheme (http: or https:)
1. len(repoParts) = 0

Either way, I don't think we should keep the error message as is, saying that it has to be two or three segments. I'm not sure what the best error message is, though.

Comment (legionus) [inline pkg/image/api/helper.go]: @ncdc This error message describes the common requirements for the DockerImageReference. It seems to me that it is better to leave it as is. Otherwise we need to add much more checks to make it better.

Comment (ncdc) [inline pkg/image/api/helper.go]: You can leave it for now, to get this PR in, but I would like it fixed when we refactor parsing.

File: test/integration/imageimporter_test.go (hunk 19 above)
Comment (miminar) [inline test/integration/imageimporter_test.go]: Thanks!

**Comment by miminar** [comment]
LGTM

File: pkg/image/api/helper_test.go [hunk 25]
```diff
@@ -291,7 +291,21 @@ func TestParseDockerImageReference(t *testing.T) {
...
+			From: "bar/foo/baz////biz",
+			Err:  true,
+		},
+		{
+			From: "//foo/baz/biz",
+			Err:  true,
+		},
+		{
+			From: "docker.io/library/foo/myapp",
 			Err:  true,
```
Comment (miminar) [inline pkg/image/api/helper_test.go]: This should be a non-error. Confirmed by jenkins:

```
--- FAIL: TestParseDockerImageReference (0.00s)
	helper_test.go:328: docker.io/library/foo/myapp: unexpected non-error
```

**Comment by ncdc** [comment]
I'm pretty sure that should error

File: pkg/image/api/helper_test.go (hunk 25 above)
Comment (legionus) [inline pkg/image/api/helper_test.go]: Sure.

**Comment by ncdc** [comment]
Heh, forgot that was removed

**Comment by legionus** [comment]
@ncdc fixed and tested.
Looks good now ?

**Comment by ncdc** [comment]
Good enough... can't wait to see this refactored though.

**Comment by mfojtik** [comment]
[merge]

(bumped priority on follow up)

**Comment by miminar** [comment]
Flake #11240 in the merge job https://ci.openshift.redhat.com/jenkins/job/test_pr_origin/9696/.

**Comment by mfojtik** [comment]
[merge]
